vmm/src/vstate.rs

```diff
@@ -275,13 +290,139 @@ impl Vcpu {
+                            self.log_boot_time();
+                        }
+                        io_bus.write(u64::from(addr), data);
+                        METRICS.vcpu.exit_io_out.inc();
+                    }
+                    VcpuExit::MmioRead(addr, data) => {
+                        mmio_bus.read(addr, data);
+                        METRICS.vcpu.exit_mmio_read.inc();
+                    }
+                    VcpuExit::MmioWrite(addr, data) => {
+                        mmio_bus.write(addr, data);
+                        METRICS.vcpu.exit_mmio_write.inc();
+                    }
+                    VcpuExit::Hlt => {
+                        info!("Received KVM_EXIT_HLT signal");
+                        break;
+                    }
+                    VcpuExit::Shutdown => {
+                        info!("Received KVM_EXIT_SHUTDOWN signal");
+                        break;
+                    }
+                    // Documentation specifies that below kvm exits are considered
+                    // errors.
+                    VcpuExit::FailEntry => {
+                        METRICS.vcpu.failures.inc();
+                        error!("Received KVM_EXIT_FAIL_ENTRY signal");
+                        break;
+                    }
+                    VcpuExit::InternalError => {
+                        METRICS.vcpu.failures.inc();
+                        error!("Received KVM_EXIT_INTERNAL_ERROR signal");
+                        break;
+                    }
+                    r => {
+                        METRICS.vcpu.failures.inc();
+                        // TODO: Are we sure we want to finish running a vcpu upon
+                        // receiving a vm exit that is not necessarily an error?
+                        error!("Unexpected exit reason on vcpu run: {:?}", r);
+                        break;
+                    }
+                },
+                // The unwrap on raw_os_error can only fail if we have a logic
+                // error in our code in which case it is better to panic.
+                Err(Error::VcpuRun(ref e)) => {
+                    match e.raw_os_error().unwrap() {
+                        // Why do we check for these if we only return EINVAL?
+                        libc::EAGAIN | libc::EINTR => {}
+                        _ => {
+                            METRICS.vcpu.failures.inc();
+                            error!("Failure during vcpu run: {:?}", e);
+                            break;
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        // Nothing we need do for the success case.
+        if let Err(e) = vcpu_exit_evt.write(1) {
+            METRICS.vcpu.failures.inc();
+            error!("Failed signaling vcpu exit event: {:?}", e);
+        }
+    }
+
+    fn run_guest(&self) -> Result<VcpuExit> {
         self.fd.run().map_err(Error::VcpuRun)
     }
+
+    fn log_boot_time(&self) {
```

Comment:
This looks more at home as a global function in the parent module (which can still be invoked in this one), especially if keep the original form of `START_INSTANCE_REQUEST_CPU_TS` and `START_INSTANCE_REQUEST_TS`.

Reply:
Makes sense. I'm still keen on not using global state if not necessary, but either way this would be more at home as a global function in the parent module. Will change.

Reply:
done